kubernetes/istio_rules_service.go

```diff
@@ -37,153 +39,97 @@ func (in *IstioClient) GetIstioRuleDetails(namespace string, istiorule string) (
-	handler := action["handler"]
-	hName, hType, hNamespace := parseFQN(handler.(string))
-	if hNamespace != "" && hNamespace != namespace {
-		return nil, fmt.Errorf("Istio Rule %s/%s has a handler for a different namespace", namespace, istiorule)
-	}
-
-	// KIALI-1819 Istio 1.1.x has introduced generic handlers
-	if hType == "" {
-		hType = handlerType
-	}
-
-	handlerChan, instancesChan := make(chan istioResponse), make(chan istioResponse)
-
-	go in.getActionHandler(namespace, hType, hName, handlerChan)
-
-	instances := action["instances"]
-	for _, instance := range instances.(instancesType) {
-		iName, iType, iNamespace := parseFQN(instance.(string))
-		go in.getActionInstance(namespace, istiorule, iNamespace, iType, iName, instancesChan)
-	}
-
-	istioRuleAction := IstioRuleAction{}
-
-	handlerResponse := <-handlerChan
-	if handlerResponse.err != nil {
-		return nil, handlerResponse.err
-	}
-	istioRuleAction.Handler = handlerResponse.result
-
-	istioRuleAction.Instances = make([]IstioObject, 0)
-	for i := 0; i < len(instances.(instancesType)); i++ {
-		instanceResponse := <-instancesChan
-		if instanceResponse.err != nil {
-			return nil, instanceResponse.err
-		}
-		istioRuleAction.Instances = append(istioRuleAction.Instances, instanceResponse.result)
-	}
+func (in *IstioClient) GetAdapter(namespace, adapterType, adapterName string) (IstioObject, error) {
+	return in.getAdapterTemplate(namespace, "adapter", adapterType, adapterName, adapterPlurals)
+}
 
-	return &istioRuleAction, nil
+func (in *IstioClient) GetTemplate(namespace, templateType, templateName string) (IstioObject, error) {
+	return in.getAdapterTemplate(namespace, "template", templateType, templateName, templatePlurals)
 }
 
-func (in *IstioClient) getActionHandler(namespace string, handlerType string, handlerName string, handlerChan chan<- istioResponse) {
-	handlerTypePlural, ok := istioTypePlurals[handlerType]
-	if !ok {
-		log.Warningf("Handler type %s is not supported", handlerType)
-		handlerChan <- istioResponse{}
-		return
-	}
-	result, err := in.istioConfigApi.Get().Namespace(namespace).Resource(handlerTypePlural).SubResource(handlerName).Do().Get()
-	istioObject, ok := result.(IstioObject)
-	if !ok {
-		istioObject = nil
-		if err == nil {
-			err = fmt.Errorf("%s/%s doesn't return a valid IstioObject", handlerType, handlerName)
+func (in *IstioClient) getAdaptersTemplates(namespace string, itemType string, pluralsMap map[string]string) ([]IstioObject, error) {
+	resultsChan := make(chan istioResponse)
+	for name, plural := range pluralsMap {
+		go func(name, plural string) {
+			results, err := in.istioConfigApi.Get().Namespace(namespace).Resource(plural).Do().Get()
+			istioObjects := istioResponse{}
+			resultList, ok := results.(*GenericIstioObjectList)
+			if !ok {
+				istioObjects.results = nil
+				istioObjects.err = fmt.Errorf("%s doesn't return a %s list", namespace, plural)
+			}
+			if err == nil {
```

Comment:
should be `else if`, no? (I guess if `!ok`, then `resultList` is going to be nil)

Reply:
Correct, I wanted to return on this "if" but I changed and introduced this bug.
Good catch.